### PR TITLE
[Tracer][ONNX] Enable numel tracing

### LINKED ISSUE #74047: tensor.numel() traced as constant
### 🐛 Describe the bug

Example for repro:
```python
import torch

class Model(torch.nn.Module):
    def forward(self, x):
        return torch.ones(x.numel())

x = torch.randn(2, 3, 4)
g = torch.jit.trace(Model(), (x, ))
print(g.graph)
```
Log:
```
graph(%self : __torch__.Model,
      %x : Float(2, 3, 4, strides=[12, 4, 1], requires_grad=0, device=cpu)):
  %4 : int = prim::Constant[value=24]() # repro_march_numel.py:5:0
  %5 : int[] = prim::ListConstruct(%4)
  %6 : int = prim::Constant[value=6]() # repro_march_numel.py:5:0
  %7 : NoneType = prim::Constant()
  %8 : Device = prim::Constant[value="cpu"]() # repro_march_numel.py:5:0
  %9 : bool = prim::Constant[value=0]() # repro_march_numel.py:5:0
  %10 : Float(24, strides=[1], requires_grad=0, device=cpu) = aten::ones(%5, %6, %7, %8, %9) # repro_march_numel.py:5:0
  return (%10)
```
Observe that `x.numel()` got traced as `%4 : int = prim::Constant[value=24]()` which has fixed value. A more accurate graph is to dynamically compute from shape of `x`, as how `tensor.size()` is traced.

### Versions

PyTorch version: 1.12.0a0+git1908433
Is debug build: False
CUDA used to build PyTorch: 11.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.3 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04) 9.4.0
Clang version: 10.0.0-4ubuntu1
CMake version: version 3.22.2
Libc version: glibc-2.17

Python version: 3.7.11 (default, Jul 27 2021, 14:32:16)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.13.0-28-generic-x86_64-with-debian-bullseye-sid
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: NVIDIA GeForce GTX 1080
Nvidia driver version: 510.47.03
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.3.2
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.3.2
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.3.2
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.3.2
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.3.2
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.3.2
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.3.2
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.21.2
[pip3] torch==1.12.0a0+git187f733

### PR DESCRIPTION
**Pull request by BowenBao**
Fixes #74047

Trace `tensor.numel()` as `aten::numel`, instead of a constant value. The logic is similar to tracing `tensor.size(dim)`.

### CODE REVIEW

File: tools/autograd/templates/python_variable_methods.cpp [hunk 1]
```diff
@@ -231,7 +231,11 @@ static PyObject * THPVariable_numel(PyObject* self, PyObject* args)
      return handle_torch_function(self, "numel", args);
    }
    auto& self_ = THPVariable_Unpack(self);
-   return THPUtils_packInt64(self_.numel());
+   if (jit::tracer::isTracing()) {
```
Comment (albanD) [inline tools/autograd/templates/python_variable_methods.cpp:234]: Why does this needs to be special case. This is not something we usually do for any other op right?

Comment (BowenBao) [reply tools/autograd/templates/python_variable_methods.cpp:234]: After reading similar snippet for `size()` (and https://github.com/pytorch/pytorch/pull/6554#issuecomment-380927254), my understanding is that operators that produce primitive typed outputs cannot be traced. Hence the special case of changing to return pytorch scalar instead of python scalar under tracing mode.

**Comment by BowenBao** [comment]
Closing as replaced by #74081. Not familiar with ghstack as I thought it will continue to build on top of this existing PR.
